widget/tree.go

```diff
@@ -0,0 +1,792 @@
+	t.walk(t.Root, 0, onNode)
+}
+
+func (t *Tree) walk(uid string, depth int, onNode func(string, bool, int)) {
+	if isBranch := t.IsBranch; isBranch != nil {
+		if isBranch(uid) {
+			onNode(uid, true, depth)
+			if t.IsBranchOpen(uid) {
+				if children := t.Children; children != nil {
+					for _, c := range children(uid) {
+						t.walk(c, depth+1, onNode)
+					}
+				}
+			}
+		} else {
+			onNode(uid, false, depth)
+		}
+	}
+}
+
+func (t *Tree) ensureOpenMap() {
+	t.propertyLock.Lock()
+	defer t.propertyLock.Unlock()
+	if t.open == nil {
+		t.open = make(map[string]bool)
+	}
+}
+
+// CreateRenderer is a private method to Fyne which links this widget to its renderer.
+func (t *Tree) CreateRenderer() fyne.WidgetRenderer {
+	t.ExtendBaseWidget(t)
+	c := newTreeContent(t)
+	s := NewScrollContainer(c)
+	r := &treeRenderer{
+		BaseRenderer: widget.NewBaseRenderer([]fyne.CanvasObject{s}),
+		tree:         t,
+		content:      c,
+		scroller:     s,
+	}
+	s.onOffsetChanged = func() {
+		if t.Offset == s.Offset {
+			return
+		}
+		t.Offset = s.Offset
+		c.Refresh()
+	}
+	r.updateMinSizes()
+	r.content.viewport = r.MinSize()
+	return r
+}
+
+var _ fyne.WidgetRenderer = (*treeRenderer)(nil)
+
+type treeRenderer struct {
+	widget.BaseRenderer
+	tree     *Tree
+	content  *treeContent
+	scroller *ScrollContainer
+}
+
+func (r *treeRenderer) MinSize() (min fyne.Size) {
+	min = r.scroller.MinSize()
+	min = min.Max(r.tree.branchMinSize)
+	min = min.Max(r.tree.leafMinSize)
+	return
+}
+
+func (r *treeRenderer) Layout(size fyne.Size) {
+	r.content.viewport = size
+	r.scroller.Resize(size)
+}
+
+func (r *treeRenderer) Refresh() {
+	r.updateMinSizes()
+	s := r.tree.Size()
+	if s.IsZero() {
+		r.tree.Resize(r.tree.MinSize())
+	}
+	r.scroller.Refresh()
+	r.content.Refresh()
+	canvas.Refresh(r.tree.super())
+}
+
+func (r *treeRenderer) updateMinSizes() {
+	if f := r.tree.CreateNode; f != nil {
+		r.tree.branchMinSize = newBranch(r.tree, f(true)).MinSize()
+		r.tree.leafMinSize = newLeaf(r.tree, f(false)).MinSize()
+	}
+}
+
+var _ fyne.Widget = (*treeContent)(nil)
+
+type treeContent struct {
+	BaseWidget
+	tree     *Tree
+	viewport fyne.Size
+}
+
+func newTreeContent(tree *Tree) (c *treeContent) {
+	c = &treeContent{
+		tree: tree,
+	}
+	c.ExtendBaseWidget(c)
+	return
+}
+
+func (c *treeContent) CreateRenderer() fyne.WidgetRenderer {
+	return &treeContentRenderer{
+		BaseRenderer: widget.BaseRenderer{},
+		treeContent:  c,
+		branches:     make(map[string]*branch),
+		leaves:       make(map[string]*leaf),
+		branchPool:   &syncPool{},
+		leafPool:     &syncPool{},
+	}
+}
+
+var _ fyne.WidgetRenderer = (*treeContentRenderer)(nil)
+
+type treeContentRenderer struct {
+	widget.BaseRenderer
+	treeContent *treeContent
+	dividers    []*canvas.Rectangle
+	branches    map[string]*branch
+	leaves      map[string]*leaf
+	branchPool  pool
+	leafPool    pool
+}
+
+func (r *treeContentRenderer) MinSize() (min fyne.Size) {
+	r.treeContent.propertyLock.Lock()
```

Comment:
Does the minsize matter as it's inside the scroll container? This looks like an expensive operation so thought I would ask :)

Reply:
Not sure I follow. Doesn't the scroller need to know how big its content is?

Reply:
I think it will use the max of MinSize and Size.
Changing MinSize has impacts in the driver as it will immediately invalidate caches that would remain if it was just Size changing. (IIRC - this is related to how a window knows if it needs to expand).

Reply:
Okay, are you proposing this MinSize return 0,0?

Reply:
I wasnt't thinking that.
On reflection though I think my concerns were unfounded, we are inside a scroller so the propagation should not occur.